tests/unittests/core/test_experiment.py

```diff
@@ -288,32 +354,57 @@ def test_get_after_init_plus_hit_no_diffs(self, exp_config):
         """
         exp = Experiment('supernaedo2')
         # Deliver an external configuration to finalize init
-        exp.configure(exp_config[0][1])
+        experiment_count_before = exp._db.count("experiments")
+        exp.configure(exp_config[0][0])
         assert exp._init_done is True
-        exp_config[0][1]['status'] = 'pending'
-        exp_config[0][1]['algorithms']['dumbalgo']['done'] = False
-        exp_config[0][1]['algorithms']['dumbalgo']['judgement'] = None
-        exp_config[0][1]['algorithms']['dumbalgo']['scoring'] = 0
-        exp_config[0][1]['algorithms']['dumbalgo']['suspend'] = False
-        exp_config[0][1]['algorithms']['dumbalgo']['value'] = 5
-        assert exp._id == exp_config[0][1].pop('_id')
-        assert exp.configuration == exp_config[0][1]
+        exp_config[0][0]['status'] = 'pending'
+        exp_config[0][0]['algorithms']['dumbalgo']['done'] = False
+        exp_config[0][0]['algorithms']['dumbalgo']['judgement'] = None
+        exp_config[0][0]['algorithms']['dumbalgo']['scoring'] = 0
+        exp_config[0][0]['algorithms']['dumbalgo']['suspend'] = False
+        exp_config[0][0]['algorithms']['dumbalgo']['value'] = 5
+        assert exp._id == exp_config[0][0].pop('_id')
+        assert exp.configuration == exp_config[0][0]
+        assert experiment_count_before == exp._db.count("experiments")
 
     def test_try_set_after_init(self, exp_config):
         """Cannot set a configuration after init (currently)."""
         exp = Experiment('supernaedo2')
         # Deliver an external configuration to finalize init
-        exp.configure(exp_config[0][1])
+        exp.configure(exp_config[0][0])
         assert exp._init_done is True
         with pytest.raises(RuntimeError) as exc_info:
-            exp.configure(exp_config[0][1])
+            exp.configure(exp_config[0][0])
         assert 'cannot reset' in str(exc_info.value)
 
+    def test_try_set_after_race_condition(self, exp_config, new_config):
```

Comment:
So this test demonstrates that if two Experiment perceive that `is_new is True`, then only one of them will actually configure and that the other one will raise a DuplicateError. The exception will get caught by `metaopt.cli.create_experiment` and creation will be retries. All second attempts for creation will succeed (assuming no forking conditions) because the second time `is_new is False`! Is this correct?

Reply:
Yes! Quite accuracy. If I may say, all experiment racing will configure, but only one will be able to write its configuration to DB, while the others will hit a DuplicateKeyError. My intention was to test the retry inside the `cli` unittests, but maybe we should also test it at a lower level on `Experiment` directly, what do you think?

Reply:
and mark it xfail? why not? it will help polishing UI details later :stuck_out_tongue:

Reply:
But it won't fail. I've added it in 4bde5c9

Reply:
@bouthilx last reply was meant for #57...

I think that even if the first one will only write a new configuration with unique (exp_name, user_name), all the rest will eventually enter the branch where `is_new is False` and they will still call `_db.write(...)` but they are going to update it to be exactly the same thing.

As these two are the only options:
```python
        if is_new:
            self._db.write('experiments', final_config)
            # XXX: Reminder for future DB implementations:
            # MongoDB, updates an inserted dict with _id, so should you :P
            self._id = final_config['_id']
        else:
            self._db.write('experiments', final_config, {'_id': self._id})
```

It could verified by a mock at `_db.write` and testing against the order of the calls to that function and which arguments they call it with. Perhaps, it could be an informative test. I am not sure..

Reply:
The `else`-clause will not update the document to exactly the same thing. Some changes do not trigger a fork, `max_trials` for instance. I would verify such modifications. Don't we have a unit test doing that?